### ISSUE TITLE
Cannot Set PlaybackRate in Windows build

### ISSUE THREAD

**Issue by CSTobler777** [RenderHeads/UnityPlugin-AVProVideo#662]
Hi, we are using your plugin in our game, it is working perfectly in a Mac build but I am having issues in a Windows build. When I call myVideo.Control.SetPlaybackRate(speed), it does not seem to set the playback rate in a windows build, whereas this works in a Mac build. The video will just stand still. Is there something I have to write differently for Windows?

Thank you,

Christian

**Comment by AndrewRH** [comment]
Hi,

If you're using version 2.x then you set playback rate by:

myVidoe.PlaybackRate = speed;

Does that solve you issue?

Thanks,

**Comment by CSTobler777** [comment]
Thank you for getting back so quickly! I changed it to the code you wrote, but it is still not working. When I set the playback rate, the video just sits, but in editor or on a Mac build, it plays normally. The same code is being used for both. Is there something different I am missing with how Windows needs to be set up?

Thanks.

**Comment by CSTobler777** [comment]
I wanted add some details. We are using unity 2019.4.18f1, The video files are in the streaming assets folder, and they are h.264 codec. The following is the pertinent block of code, this gets called every frame. The different points are where the video slows, and then stops, and the last else statement is where the normal playback speed is called. I have debugged it and confirmed that it is reaching this line of code, and the value it is trying to set is correct. 

`if (Mathf.Approximately(Mathf.Clamp((float)videoPlayerUp.Control.GetCurrentTime(), ITVLogic.point1, ITVLogic.point2 - 0.1f), (float)videoPlayerUp.Control.GetCurrentTime()) ||
            Mathf.Approximately(Mathf.Clamp((float)videoPlayerUp.Control.GetCurrentTime(), ITVLogic.point3, ITVLogic.point4 - 0.1f), (float)videoPlayerUp.Control.GetCurrentTime()) ||
            Mathf.Approximately(Mathf.Clamp((float)videoPlayerUp.Control.GetCurrentTime(), ITVLogic.point5, ITVLogic.point6 - 0.1f), (float)videoPlayerUp.Control.GetCurrentTime()) ||
            Mathf.Approximately(Mathf.Clamp((float)videoPlayerUp.Control.GetCurrentTime(), ITVLogic.point7, ITVLogic.point8 - 0.1f), (float)videoPlayerUp.Control.GetCurrentTime()))
            {
                videoPlayerUp.PlaybackRate = 0;
                Debug.LogError("Stopped in First(at point)");
                //videoPlayerDN.playbackSpeed = 0;
            }
            else if (Mathf.Approximately(Mathf.Clamp((float)videoPlayerUp.Control.GetCurrentTime(), ITVLogic.point1 - .5f, ITVLogic.point1), (float)videoPlayerUp.Control.GetCurrentTime())) // the .5f number is the point where it starts to slow down. used to be 1, for the longer vid.
            {
                Debug.LogError("Slowing down at point 1");
                float t = Mathf.InverseLerp(ITVLogic.point1 - .5f, ITVLogic.point1, (float)videoPlayerUp.Control.GetCurrentTime());
                float speed = Mathf.Lerp(.5f, slowestSpeed, t);
                videoPlayerUp.PlaybackRate = speed;
            }
            else if (Mathf.Approximately(Mathf.Clamp((float)videoPlayerUp.Control.GetCurrentTime(), ITVLogic.point3 - .5f, ITVLogic.point3), (float)videoPlayerUp.Control.GetCurrentTime()))
            {
                Debug.LogError("Slowing down at point 3");
                float t = Mathf.InverseLerp(ITVLogic.point3 - .5f, ITVLogic.point3, (float)videoPlayerUp.Control.GetCurrentTime());
                float speed = Mathf.Lerp(.5f, slowestSpeed, t);
                videoPlayerUp.PlaybackRate = speed;
            }
            else if (Mathf.Approximately(Mathf.Clamp((float)videoPlayerUp.Control.GetCurrentTime(), ITVLogic.point5 - .5f, ITVLogic.point5), (float)videoPlayerUp.Control.GetCurrentTime()))
            {
                Debug.LogError("Slowing down at point 5");
                float t = Mathf.InverseLerp(ITVLogic.point5 - .5f, ITVLogic.point5, (float)videoPlayerUp.Control.GetCurrentTime());
                float speed = Mathf.Lerp(.5f, slowestSpeed, t);
                videoPlayerUp.PlaybackRate = speed;
            }
            else if (Mathf.Approximately(Mathf.Clamp((float)videoPlayerUp.Control.GetCurrentTime(), ITVLogic.point7 - .5f, ITVLogic.point7), (float)videoPlayerUp.Control.GetCurrentTime()))
            {
                Debug.LogError("Slowing down at point 7");
                float t = Mathf.InverseLerp(ITVLogic.point7 - .5f, ITVLogic.point7, (float)videoPlayerUp.Control.GetCurrentTime());
                float speed = Mathf.Lerp(.5f, slowestSpeed, t);
                videoPlayerUp.PlaybackRate = speed;

            }
            else if ((float)videoPlayerUp.Control.GetCurrentTime() >= ITVLogic.point9)
            {
                Debug.LogError("Stopped in end reached");
                videoPlayerUp.PlaybackRate = 0;
                EndReached();
            }
            else
            {
                Debug.LogError("Entered into else(should be running forward.");
                Debug.LogError("Video Speed should be: " + RobotSpeed(true));// it debugs to this point, and the speed value is correct but the video does not play.
                videoPlayerUp.PlaybackRate = RobotSpeed(true);

                //videoPlayerDN.playbackSpeed = 1;
            }`

Thank you!

**Comment by kahnivore** [comment]
This is probably due to the video decoder - some decoders handle changing playback rate better than others, and it's not really meant to be called every frame. 

You could potentially get around it by encoding the video to be very simple i.e. keyframe only - which would make it easier for the decoder to deal with. 

Alternatively if you don't have audio, it'd be better to pause the video and adjust it manually.

**Comment by CSTobler777** [comment]
Ok. We are not using audio, so you recommend pausing the video and then seeking to simulate the video playing? I will try that next. Additionally, do you support ogg format? And would you recommend it in this situation?

Thanks

**Comment by CSTobler777** [comment]
Well, we encoded our video to be keyframed every frame, and still it doesn't work. I did some more testing, we set the PlaybackRate to 1 once(not in any update loop) and the video stands still. Note: I have not called Play(), or anything before I set the PlaybackRate, other than waiting for the video to finish initializing. Do I need to call play or anything else before I set the PlaybackRate for Windows? What I am doing right now works perfectly in Editor, on iOS, and on Mac. The video just won't play. 

Also, what is the difference between Media Foundation, DirectShow, and WinRT? We have been using Media foundation, should we try another API? What is your recommendation?

Thank you.
